Kernel/FileSystem/InodeWatcher.cpp

```diff
@@ -62,37 +69,90 @@ KResultOr<size_t> InodeWatcher::read(FileDescription&, u64, UserOrKernelBuffer&
     return bytes_to_write;
 }
 
-KResultOr<size_t> InodeWatcher::write(FileDescription&, u64, const UserOrKernelBuffer&, size_t)
+KResult InodeWatcher::close()
 {
-    return EIO;
+    Locker locker(m_lock);
+
+    for (auto& entry : m_wd_to_watches) {
+        auto& inode = const_cast<Inode&>(entry.value->inode);
+        inode.unregister_watcher({}, *this);
+    }
+
+    m_wd_to_watches.clear();
+    m_inode_to_watches.clear();
+    return KSuccess;
 }
 
 String InodeWatcher::absolute_path(const FileDescription&) const
 {
-    if (auto inode = m_inode.strong_ref())
-        return String::formatted("InodeWatcher:{}", inode->identifier().to_string());
-    return "InodeWatcher:(gone)";
+    return String::formatted("InodeWatcher:({})", m_wd_to_watches.size());
 }
 
-void InodeWatcher::notify_inode_event(Badge<Inode>, InodeWatcherEvent::Type event_type)
+void InodeWatcher::notify_inode_event(Badge<Inode>, InodeIdentifier inode_id, InodeWatcherEvent::Type event_type, String const& name)
 {
     Locker locker(m_lock);
-    m_queue.enqueue({ event_type });
+
+    auto it = m_inode_to_watches.find(inode_id);
+    if (it == m_inode_to_watches.end())
+        return;
+
+    auto& watcher = *it->value;
+    if (!(watcher.event_mask & static_cast<unsigned>(event_type)))
+        return;
+
+    m_queue.enqueue({ watcher.wd, event_type, name });
     evaluate_block_conditions();
 }
 
-void InodeWatcher::notify_child_added(Badge<Inode>, const InodeIdentifier& child_id)
+KResultOr<int> InodeWatcher::register_inode(Inode& inode, unsigned event_mask)
 {
     Locker locker(m_lock);
-    m_queue.enqueue({ InodeWatcherEvent::Type::ChildAdded, child_id.index().value() });
-    evaluate_block_conditions();
+
+    if (m_inode_to_watches.find(inode.identifier()) != m_inode_to_watches.end())
+        return EEXIST;
+
+    int wd = m_wd_counter.value();
+    m_wd_counter++;
+    if (m_wd_counter.has_overflow())
+        m_wd_counter = 1;
+
+    auto description = WatchDescription::create(wd, inode, event_mask);
+    m_inode_to_watches.set(inode.identifier(), description.ptr());
+    m_wd_to_watches.set(wd, move(description));
```

Comment:
If `m_wd_counter` wraps, we may now end up reusing an in-use WD number.

The old `WatchDescription` object is freed (by replacing it in `m_wd_to_watches`) and now we have a dangling pointer to the old `WatchDescription` in `m_inode_to_watches`.

We must ensure that WD numbers are only reused once they have been deallocated. The most straightforward way to do this might be to simply increment in a loop until you hit a WD number that isn't in `m_wd_to_watches`.

Reply:
You're right, I didn't think of that :^)